### ISSUE TITLE
Transformation stage fails with "Path does not exist" in DataFrameGraphReader.read()

### ISSUE THREAD

**Issue by arvindshmicrosoft** [ldbc/ldbc_snb_datagen_spark#322]
Generation stage seems okay, but the Transformation stage fails with a "path does not exist" exception. I did some analysis (see below) to hopefully trigger some clues in your minds. Hoping to get some insights from you all on this issue.

# Environment
* ldbc_snb_datagen_spark commit hash a2ffaed80dcc38307d4ffb88cf020ba927be4f8a
* Spark 2.4.8 running on Yarn
* Hadoop 3.3.1 with 8 worker nodes
* OpenJDK Runtime Environment (build 1.8.0_292-b10)

# Repro
```bash
./tools/run.py ./target/ldbc_snb_datagen_2.11_spark2.4-0.4.0-SNAPSHOT.jar --parallelism 1 -- --format csv --scale-factor 0.003 --mode interactive
```
# Analysis
The HDFS path is being pre-pended with the local working folder path, which in this case is:
```
[centos@ldbc-ldr-2 ldbc_snb_datagen_spark]$ pwd
/home/centos/ldbc_snb_datagen_spark
```
If you look at the path on HDFS without any prefix, there is actually a file here:
```
[centos@ldbc-ldr-2 ldbc_snb_datagen_spark]$ hdfs dfs -ls -R out/csv/raw/composite-merged-fk/dynamic/Forum
-rw-r--r--   3 centos supergroup      46794 2021-08-09 05:03 out/csv/raw/composite-merged-fk/dynamic/Forum/part_0_0.csv
```
But because somehow the prefixing of the local path, that is causing the "path does not exist" problem.

When I look at the [code](https://github.com/ldbc/ldbc_snb_datagen_spark/blob/a2ffaed80dcc38307d4ffb88cf020ba927be4f8a/src/main/scala/ldbc/snb/datagen/transformation/io/GraphReader.scala#L59) and debugged, the individual elements seem okay:
variable | value
------|-------
path |out
options.format | csv
PathComponent[GraphLike[M]].path(definition) | raw/composite-merged-fk
entity.entityPath | dynamic/Forum

<s>My suspicion is something is wrong in the line which actually constructs the path to pass to `.load()`, and somehow in that process, the toString() or something is causing the local folder path to get prefixed.</s>
After a bit more research, the issue is indeed due to these `better-files` idiomatic expression of path concatenation, wherein it is automatically prefixing the local folder path.
```scala
(path / options.format / PathComponent[GraphLike[M]].path(definition) / entity.entityPath).toString()
```

# Error details
```txt
2021-08-09 05:03:39,352 INFO scheduler.DAGScheduler: Job 5 finished: foreachPartition at SparkPersonSerializer.scala:25, took 0.364199 s
2021-08-09 05:03:39,618 INFO internal.SharedState: Setting hive.metastore.warehouse.dir ('null') to the value of spark.sql.warehouse.dir ('file:/home/centos/ldbc_snb_datagen_spark/spark-warehouse').
2021-08-09 05:03:39,619 INFO internal.SharedState: Warehouse path is 'file:/home/centos/ldbc_snb_datagen_spark/spark-warehouse'.
2021-08-09 05:03:39,625 INFO handler.ContextHandler: Started o.s.j.s.ServletContextHandler@58b311ba{/SQL,null,AVAILABLE,@Spark}
2021-08-09 05:03:39,626 INFO handler.ContextHandler: Started o.s.j.s.ServletContextHandler@482c351d{/SQL/json,null,AVAILABLE,@Spark}
2021-08-09 05:03:39,626 INFO handler.ContextHandler: Started o.s.j.s.ServletContextHandler@6af310c7{/SQL/execution,null,AVAILABLE,@Spark}
2021-08-09 05:03:39,627 INFO handler.ContextHandler: Started o.s.j.s.ServletContextHandler@4aed311e{/SQL/execution/json,null,AVAILABLE,@Spark}
2021-08-09 05:03:39,628 INFO handler.ContextHandler: Started o.s.j.s.ServletContextHandler@26c77f54{/static/sql,null,AVAILABLE,@Spark}
2021-08-09 05:03:39,966 INFO state.StateStoreCoordinatorRef: Registered StateStoreCoordinator endpoint
2021-08-09 05:03:39,971 WARN sql.SparkSession$Builder: Using an existing SparkSession; some spark core configurations may not take effect.
2021-08-09 05:03:39,979 INFO io.DataFrameGraphReader: Reading Forum
Exception in thread "main" org.apache.spark.sql.AnalysisException: Path does not exist: hdfs://mycluster/home/centos/ldbc_snb_datagen_spark/out/csv/raw/composite-merged-fk/dynamic/Forum;
        at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$org$apache$spark$sql$execution$datasources$DataSource$$checkAndGlobPathIfNecessary$1.apply(DataSource.scala:576)
        at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$org$apache$spark$sql$execution$datasources$DataSource$$checkAndGlobPathIfNecessary$1.apply(DataSource.scala:559)
        at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
        at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:241)
        at scala.collection.immutable.List.flatMap(List.scala:355)
        at org.apache.spark.sql.execution.datasources.DataSource.org$apache$spark$sql$execution$datasources$DataSource$$checkAndGlobPathIfNecessary(DataSource.scala:559)
        at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:373)
        at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:242)
        at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:230)
        at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:197)
        at ldbc.snb.datagen.transformation.io.DataFrameGraphReader$$anonfun$2.apply(GraphReader.scala:59)
        at ldbc.snb.datagen.transformation.io.DataFrameGraphReader$$anonfun$2.apply(GraphReader.scala:55)
        at scala.collection.TraversableLike$WithFilter$$anonfun$map$2.apply(TraversableLike.scala:683)
        at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221)
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:682)
        at ldbc.snb.datagen.transformation.io.DataFrameGraphReader.read(GraphReader.scala:55)
        at ldbc.snb.datagen.transformation.TransformationStage$.run(TransformationStage.scala:108)
        at ldbc.snb.datagen.spark.LdbcDatagen$.run(LdbcDatagen.scala:138)
        ... (call stack truncated for readability)
```

**Comment by dszakallas** [comment]
Thanks for reporting this issue. Indeed, this seems to be caused by the way better-files handles relative paths.

**Comment by arvindshmicrosoft** [comment]
@dszakallas you seem to have read my mind! I took a quick spin with your commit c6269347e9ade44005644cd4424b0d9216c4b415 and indeed it not only fixed this current issue, but also my underlying real issue which was to get it to output directly to ADLS Gen2. I tested with both local (`run.py`) and cluster (`spark-submit`) modes of operation, and each time also specifying an `--output-dir` directly to ADLS Gen2.

Hoping your commit merges with main soon. Thank you again for your quick help!
